src/storefront.rs

```diff
@@ -0,0 +1,371 @@
+                components[0].tag()
+            )
+            .into());
+        }
+
+        let component = components[0];
+
+        /* Make sure the id tag is correct */
+        let id_tags: Vec<_> = component.find_all("id").collect();
+        if id_tags.len() > 1 {
+            return Err("Duplicate <id/> tag in appstream file".into());
+        } else if id_tags.is_empty() {
+            return Err("No <id/> tag found in appstream file".into());
+        } else if id_tags[0].text() != expected_appid {
+            return Err(format!(
+                "Expected app ID to be '{}', not '{}'",
+                expected_appid,
+                id_tags[0].text()
+            )
+            .into());
+        }
+
+        for element in component.children() {
+            match element.tag().name() {
+                "url" => {
+                    /* Filter existing URLs */
+                    if let Some(url_type) = element.get_attr("type") {
+                        if !self.check_xml_tags.check(url_type) {
+                            return Err(format!(
+                                "Appstream <url type='{}'/> is not permitted.",
+                                url_type
+                            )
+                            .into());
+                        }
+                    } else {
+                        return Err("Appstream <url/> must have a 'type' attribute.".into());
+                    }
+                }
+
+                "custom" => {
+                    for custom_child in element.children() {
+                        if let Some(key) = custom_child.get_attr("key") {
+                            if !self.check_custom_values.check(key) {
+                                return Err(format!(
+                                    "Appstream <value key='{}'/> is not permitted.",
+                                    key
+                                )
+                                .into());
+                            }
+                        } else {
+                            return Err("Appstream <value/> must have a 'key' attribute.".into());
+                        }
+                    }
+                }
+
+                other => {
+                    if !self.check_xml_tags.check(other) {
+                        return Err(format!("Appstream tag <{}/> is not permitted.", other).into());
+                    }
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    pub fn apply_storefront_info(
+        &self,
+        original_appstream: &str,
+    ) -> Result<String, Box<dyn Error>> {
+        let mut root = Element::from_reader(original_appstream.as_bytes())?;
+
+        let mut components: Vec<_> = root.children_mut().collect();
+        let component = &mut components[0];
+
+        fn find_element<'a>(
+            parent: &'a mut Element,
+            tag: &'a str,
+            attr: Option<(&'a str, &'a str)>,
+        ) -> Option<&'a mut Element> {
```

Comment:
This is an awful lot of lifetimes tied together, is this neccessary/intentional?  If not, eliding the `'a` from anywhere you did not mean to tie to the output borrow will give the compiler a little more leeway.  I'm guessing that in reality the only lifetime tie should be between the parent and the return value.

This comment similarly applies to the next function down.

Reply:
I can elide the `'a` from `attr` in both, but not from `tag` due to the signature of `find_all_mut`.